articles/ensuring-acid-compliance-in-database-transactions/index.md

```diff
@@ -0,0 +1,67 @@
+
+For instance, say that you want to withdraw $70 and your relative $50 from the same account with a balance of $200. Either you will be required to wait until your relative&#39;s transaction of $50 is complete (you will then withdraw from the remaining balance of $150). Or your relative will have to wait for your transaction of $70 to complete (your relative will then withdraw from the remaining balance of 130).
+
+#### Durability
+
+Durability guarantees that completed transactions survive permanently, even in instances of system failure. This way, data within the database remains uncorrupted after a service crash, outage, or other failures. Every time portions of the database or the entire database are restarted, the system references [changelogs](https://en.wikipedia.org/wiki/Changelog#) to ensure transaction durability.
+
+Your relative can withdraw $50 only after your transaction is completed and updated in the database in the above example. Assuming that the system fails before your transaction is logged in the database, your relative cannot withdraw any money, and the account will return to its initial consistent state.
+
+### How ACID Works in Practice
+
+There are various techniques used to enforce ACID compliance, as discussed below.
+
+The [write-ahead logging](https://www.postgresql.org/docs/9.1/wal-intro.html) (WAL) technique enforces integrity, durability, and atomicity by ensuring that all transaction details are first written to a log that includes undo and redo information. In case of a transaction failure, the database checks the log and compares its contents to the database&#39;s state before the transaction failure.
+
+Imagine your computer loses power while you are in the middle of performing a transaction. When you restart your machine, the program will need to know the state of the operation before your computer went off unexpectedly, that is, whether it failed, succeeded, or succeeded partially. The write-ahead log comes to ensure that the program checks this log and compares what it was expected to perform when it suddenly lost power to what it was able to execute. On this basis, the program may keep things as they are, complete what it had started, or undo what it had started.
+
+Another technique to ensure ACID compliance is [shadow paging](https://www.geeksforgeeks.org/shadow-paging-dbms/). This [copy-on-write](https://www.computerhope.com/jargon/c/copy-on-write.htm#) technique avoids in-place page updates and instead allocates a shadow page when a page is modified. You can modify this shadow page liberally without minding consistency constraints because it has no references from other pages on disk. Pages that referred to the original page can only be updated to refer to the new page once the shadow page is ready to become durable. The page is considered automatic because it is activated only when it is ready.
```

Comment:
What does liberally mean?   You can modify this shadow page liberally? Check this once.

Reply:
Can you rewrite that sentence?